content/js1/sprints/2/prep/index.md

```diff
@@ -8,4 +8,406 @@ backlog= 'Module-JS1'
+- Give examples of how functionality can be generalised
+
+## Notes
+
+### 🧩 Percentages
+
+Given a number I want to convert it into a percentage format.
+
+For example, given the number `0.5` we should get `"50%"`
+And given the number `0.23` we should get `"23%"` and so forth.
+
+### Restating the problem
+
+We want to implement some functionality to convert _any_ given number into a percentage.
+Previously, we saw how to use some functions to perform a given task. Here are some examples:
+
+```js {linenos=table,linenostart=1}
+console.log("hello world"); // prints "hello world" to the console
+Math.round(3.141); // will output 3
+```
+
+All these expressions are _function calls_: we’re passing input to the functions, `console.log` or `Math.round` to use their functionality. In the case of `Math.round` and `console.log` these functions have been created and made available to use whenever we start running a program with NodeJS.
+
+However, there is so no such pre-built function that allows us to convert any number to a percentage. So we're going to create a function called `convertToPercentage` with the following properties:
+
+- takes any number input
+- converts it to a string output in percentage form.
+
+```js {linenos=table,linenostart=1}
+convertToPercentage(0.5); // should return "50%"
+```
+
+```js {linenos=table,linenostart=1}
+convertToPercentage(0.23); // should return "23%"
+```
+
+### Useful expressions
+
+It is often helpful to solve a problem in one specific instance before doing it for all cases.
+
+In programming, we always try the simplest thing first. Let’s consider how to convert just one number in to a percentage.
+Let’s consider the variable declaration below:
+
+```js {linenos=table,linenostart=1}
+const num = 0.5;
+```
+
+We want to create an expression for the percentage using the value of `num`.
+To convert to a percentage, we can multiply the number by 100 and then add a `"%"` sign on the end.
+
+So we can write:
+
+```js {linenos=table,linenostart=1}
+const num = 0.5;
+const percentage = `${num * 100}%`;
+```
+
+Recalling template literals, the expression in the curly braces will be evaluated first and then inserted into the string, giving us the percentage string.
+
+{{<note type="exercise" title="Exercise 1.1">}}
+
+The following variables represent the dimensions of a rectangle:
+
+```js
+height; // could have a value like 10
+width; // could have a value like 30
+```
+
+Write expressions for the following in terms of the variables above:
+a) area of the rectangle
+b) perimeter of the rectangle
+
+{{</note>}}
+
+{{<note type="exercise" title="Exercise 1.2">}}
+
+The following variable represents the cost of something in pence:
+
+```js
+price; // could have a value like "130p"
+```
+
+Write an expression for the price in pounds. The price in pounds should start with a "£" symbol. The price should be to 2 decimal places.
+
+{{</note>}}
+
+### ♻️ Wrapping up functionality
```

Comment:
I have the feeling that we are jumping into defining functions without having properly shown what a program is (because all we've seen up to now are one-liners). I think it would be valuable to explain/show that (1) a program is a set of instructions that are executed in sequence, and that (2) we go through a series of intermediate states until (3) we reach the desired result. That could help reinforce the idea of why functions are a useful tool.

I think the code visualizer is an awesome tool for understanding what happens in a program and embedding it in the content is a brilliant idea. What if, before jumping into functions, we show a simple program running in the code visualizer?

For instance:
```javascript
// Let's calculate what an amount of seconds is in days, hours, minutes and seconds
const totalSeconds = 340368;

const remainingSeconds = totalSeconds % 60;
const totalMinutes = (totalSeconds - remainingSeconds) / 60;
const remainingMinutes = totalMinutes % 60;
const totalHours = (totalMinutes - remainingMinutes) / 60;
const remainingHours = totalHours % 24;
const days = (totalHours - remainingHours) / 24;

const result = `days: ${days}, hours: ${remainingHours}, minutes: ${remainingMinutes}, seconds: ${remainingSeconds}`;
console.log(result);
```
By running step by step through this program with the code visualizer, the trainee will see as intermediate results are being calculated and stored to, in sequence, reach the desired result. This is just an example, maybe we could come up with a simpler piece of code, because we don't necessarily want the trainee trying to understand this concrete algorithm, we just want the trainee to understand what a program is, and how it's executed.

This change would require moving the "Current line of code" and "Storing things in memory" sections up, though.